nixos/modules/programs/fish.nix

```diff
@@ -105,72 +129,165 @@ in
-      and begin
-        set --prepend fish_function_path ${pkgs.fishPlugins.foreign-env}/share/fish/vendor_functions.d
-        fenv source /etc/fish/foreign-env/loginShellInit > /dev/null
-        set -e fish_function_path[1]
-
-        ${cfg.loginShellInit}
-
-        # and leave a note so we don't source this config section again from
-        # this very shell (children will source the general config anew)
-        set -g __fish_nixos_login_config_sourced 1
-      end
-
-      # if we haven't sourced the interactive config, do it
-      status --is-interactive; and not set -q __fish_nixos_interactive_config_sourced
-      and begin
-        ${fishAliases}
-
-        set --prepend fish_function_path ${pkgs.fishPlugins.foreign-env}/share/fish/vendor_functions.d
-        fenv source /etc/fish/foreign-env/interactiveShellInit > /dev/null
-        set -e fish_function_path[1]
-
-        ${cfg.promptInit}
-        ${cfg.interactiveShellInit}
-
-        # and leave a note so we don't source this config section again from
-        # this very shell (children will source the general config anew,
-        # allowing configuration changes in, e.g, aliases, to propagate)
-        set -g __fish_nixos_interactive_config_sourced 1
-      end
-    '';
+    environment = mkMerge [
+      (mkIf cfg.translateEnvironment
+      {
+        etc."fish/set-environment.fish".source = pkgs.runCommand "set-environment.fish" {
+        nativeBuildInputs = [ pkgs.babelfish ];
+      } ''
+        ${pkgs.babelfish}/bin/babelfish < ${config.system.build.setEnvironment} > $out;
+      '';
+
+       etc."fish/shellInit".source = pkgs.runCommand "shellInit" {
+        nativeBuildInputs = [ pkgs.babelfish ];
+      } ''
+        ${pkgs.babelfish}/bin/babelfish < ${envShellInit} > $out;
+      '';
+
+       etc."fish/loginShellInit".source = pkgs.runCommand "loginShellInit" {
+        nativeBuildInputs = [ pkgs.babelfish ];
+      } ''
+        ${pkgs.babelfish}/bin/babelfish < ${envLoginShellInit} > $out;
+      '';
+
+       etc."fish/interactiveShellInit".source = pkgs.runCommand "interactiveShellInit" {
+        nativeBuildInputs = [ pkgs.babelfish ];
+      } ''
+        ${pkgs.babelfish}/bin/babelfish < ${envInteractiveShellInit} > $out;
+      '';
+     })
+
+      (mkIf (!cfg.translateEnvironment)
+      {
+        etc."fish/shellInit".text = cfge.shellInit;
+        etc."fish/loginShellInit".text = cfge.loginShellInit;
+        etc."fish/interactiveShellInit".text = cfge.interactiveShellInit;
+      })
+
+      {
+        etc."fish/nixos-env-preinit.fish".text = ''
+        # This happens before $__fish_datadir/config.fish sets fish_function_path, so it is currently
+        # unset. We set it and then completely erase it, leaving its configuration to $__fish_datadir/config.fish
+        set fish_function_path ${pkgs.fishPlugins.foreign-env}/share/fish/vendor_functions.d $__fish_datadir/functions
```

Comment:
Seems weird to rely on both foreign-env and babelfish if babelfish is enabled. This should probably be gated on babelfish not being used.

Reply:
I agree. It's a little tricky because the guard below splicing the setup and teardown makes the function above not applicable, but I'll try to do it as cleanly as possible to avoid the dependency.